### ISSUE TITLE
False negative "JSC_TYPE_MISMATCH" warning

### ISSUE THREAD

**Issue by patrick96** [google/closure-compiler#3865]
Using `ADVANCED` optimizations to compile the following code leads to a wrong warning:

```js
const x = {
    foo: null
};

function bar() {
    console.log(x.foo);

    if (x.foo !== null && x.foo instanceof Object) {
        return x.foo;
    } else {
        x.foo = {};
    }
}

bar();
bar();
```

The compiler warns us that the `instanceof` is check is deterministic and always `false`:

```
instanceof.js:8:28: WARNING - [JSC_TYPE_MISMATCH] deterministic instanceof yields false
found   : null
required: NoObject
  8|     if (x.foo !== null && x.foo instanceof Object) {
                                 ^^^

0 error(s), 1 warning(s), 100.0% typed
```

This is clearly not true since `x.foo` gets set to `{}`.

**To Reproduce**

I compiled using the following command:

```
npx google-closure-compiler --js instanceof.js --js_output_file tmp.js -O ADVANCED
```

**Additional Context**

Annotating the `foo` property with `/** @type {?Object} */` makes the warning disappear.

Doing a `x.foo != null` check instead of a `!==` check avoids this warning but leads to a different wrong warning:

```
instanceof.js:8:8: WARNING - [JSC_DETERMINISTIC_TEST] condition always evaluates to false
left : null
right: null
  8|     if (x.foo != null && x.foo instanceof Object) {
             ^^^^^^^^^^^^^

0 error(s), 1 warning(s), 100.0% typed
```

This seems to be the same issue as #1748 and thus isn't a viable workaround.

It also happens with a smaller example:

```js
const x = null;
console.log((x !== null && x instanceof Object));
```

**Environment**

Installed through npm:

```
$ npx google-closure-compiler --version
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20210907
```

Let me know if you need any additional context.

**Comment by concavelenz** [comment]
Yes, the compiler want you to annotate the `foo` field with the type.  It is a limitation of the inference.  The behavior you are seeing is because the type is seen as "null" at the top of the function and the type is update as it flows through the function but this not is propagated back.

That it tightens properties in the local scope is questionable, that it isn't an error assign anything other than null after the type is inferred to be null is also questionable.

But it is unfortuately, unlikely something that is going to change.

**Comment by patrick96** [comment]
I see. Thanks for the explanation. Fortunately, the workaround is easy enough.
